Dev catalog- group by Operator

debsmita1 opened the pull request:
This PR includes:
- adds a `Group By` dropdown, in the Developer Catalog, with `None` and `Operator` as options
- when Group By `Operator` is selected then all the Installed Operators which are owned by a CSV are grouped together, and all the Non-Operators are grouped together separately

Refer JIRA story: https://jira.coreos.com/browse/ODC-2452

![Screenshot from 2019-12-20 19-44-21](https://user-images.githubusercontent.com/22490998/71261487-7fbc4200-2363-11ea-8efc-f771e421b766.png)


Watch video here: https://youtu.be/lbXgLkb05uo

ToDo:
- waiting on UX for the final suggestions
- handle the scenario where Group By Operator and the categories on top-left of the Dev Catalog are selected

debsmita1 commented:
^ @openshift/team-devconsole-ux

serenamarie125 commented:
Summary of the points we discussed in our meeting:

- The order of the items in the Group by Operator should be changed. Operator should be first.
- For a "catch all" grouping, for now use Non Operators. We can modify this at a later date if a better term is found
- A single grey background should encompass all of the groupings, including the group by titles and tiles
- Decrease the amount of space after the group by titles by 50%, this will help to create a stronger visual connection between the group title and its tiles
- Move the Group by drop down closer to the filter component. Try using 32px (xl spacer) between the two components.
- Once you make these changes, please share again. Currently the group by titles stand out a bit too much. Having them placed on the grey background may help. If not, I'd suggest making them regular font rather than bold.

debsmita1 commented:
> Summary of the points we discussed in our meeting:
> 
>     * The order of the items in the Group by Operator should be changed. Operator should be first.
Done 
>     * For a "catch all" grouping, for now use Non Operators. We can modify this at a later date if a better term is found
Done
> 
>     * A single grey background should encompass all of the groupings, including the group by titles and tiles
> 
>     * Decrease the amount of space after the group by titles by 50%, this will help to create a stronger visual connection between the group title and its tiles
Added `5px` space between the title and the group
> 
>     * Move the Group by drop down closer to the filter component. Try using 32px (xl spacer) between the two components.
Done. Using `--pf-global--spacer--xl`
> 
>     * Once you make these changes, please share again. Currently the group by titles stand out a bit too much. Having them placed on the grey background may help. If not, I'd suggest making them regular font rather than bold.
Made the font regular. Using `--pf-global--FontSize--lg`

@serenamarie125 I have added a youtube link in the PR description. Please let me know if it looks good to you
cc @christianvogt

debsmita1 commented:
/cc @christianvogt 
/cc @abhinandan13jan

rohitkrai03 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -21,9 +21,18 @@ import {
 
-const CATEGORY_URL_PARAM = 'category';
-const KEYWORD_URL_PARAM = 'keyword';
+export const FilterType = {
```
Use enum here.

rohitkrai03 replied:
Better name it `FilterTypes`.

debsmita1 replied:
used const instead of enum because it's giving an error saying `'enum declarations' can only be used in a .ts file` cc @christianvogt

christianvogt replied:
oh haha... jsx... didn't pick up on that one :)
Time for an upgrade!

rohitkrai03 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -21,9 +21,18 @@ import {
+  groupBy: 'groupBy',
+};
+
+export const groupByTypes = {
```
Same as above.

rohitkrai03 replied:
A better name for the enum would be `GroupTypes`.

christianvogt replied:
Why not an enum?

rohitkrai03 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -21,9 +21,18 @@ import {
+export const FilterType = {
+  category: 'category',
+  keyword: 'keyword',
+  groupBy: 'groupBy',
```
No need to use `groupBy` everywhere.
```suggestion
  group: 'group',
```

christianvogt replied:
Grouping should not be a filter.

rohitkrai03 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -227,6 +235,17 @@ const filterByGroup = (items, filters) => {
   );
 };
 
+export const filterByGroupByDropdown = (items, groupBy) => {
```
```suggestion
export const filterByGroup = (items, group) => {
```

christianvogt replied:
It's not really a filter though. It's a sort or grouping. I'd remove filter from this.

debsmita1 replied:
Changed it to `groupItems`

rohitkrai03 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -227,6 +235,17 @@ const filterByGroup = (items, filters) => {
+  if (groupBy === groupByTypes.Operator) {
+    const installedOperators = _.filter(items, (item) => item.kind === 'InstalledOperator');
+    const nonOperators = _.filter(items, (item) => item.kind !== 'InstalledOperator');
+    const groupedOperators = _.groupBy(installedOperators, (item) => item.obj.csv.metadata.name);
```
Maybe we can sanitize the operator name somehow to make it look better on the UI. The title currently looks really odd.

debsmita1 replied:
@rohitkrai03 Do you think the version is making it look odd? cc @serenamarie125

debsmita1 replied:
@rohitkrai03  Let me know if this looks okay 
![Screenshot from 2019-12-19 23-40-18](https://user-images.githubusercontent.com/22490998/71198055-fb14e980-22b8-11ea-9f46-82fe9250fb4f.png)

rohitkrai03 replied:
Yeah, this looks much better.

debsmita1 replied:
Done

rohitkrai03 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -227,6 +235,17 @@ const filterByGroup = (items, filters) => {
+    const installedOperators = _.filter(items, (item) => item.kind === 'InstalledOperator');
+    const nonOperators = _.filter(items, (item) => item.kind !== 'InstalledOperator');
+    const groupedOperators = _.groupBy(installedOperators, (item) => item.obj.csv.metadata.name);
+    const groupAllItems = { ...groupedOperators, 'Non-Operators': nonOperators };
```
Why add a hyphen in between when its a string?
```suggestion
    const groupAllItems = { ...groupedOperators, 'Non Operators': nonOperators };
```

rohitkrai03 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -310,10 +329,13 @@ const determineAvailableFilters = (initialFilters, items, filterGroups) => {
 };
 
-const getActiveFilters = (keywordFilter, groupFilters, activeFilters) => {
+const getActiveFilters = (keywordFilter, groupByFilter, groupFilters, activeFilters) => {
```
Oh, i see why you had to name it `groupBy` everywhere.

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -764,8 +819,34 @@ export class TileViewPage extends React.Component {
+      (value, key) =>
+        value.length > 0 && (
+          <>
+            <div className="co-catalog-page__group-title">
```
Use a standard header? `h2` perhaps

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -764,8 +819,34 @@ export class TileViewPage extends React.Component {
+              className="catalog-tile-view-pf catalog-tile-view-pf-no-categories co-catalog-page__grid"
+            >
+              {_.map(value, (item) => (
+                <GalleryItem key={`gallery-${item.obj.metadata.uid}`}>
```
This isn't consistent with the other gallery item.
It's also duplicate code that should be extracted and reused.

debsmita1 replied:
Moved this part to `renderGalleryItem` and reusing it whereever required

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -764,8 +819,34 @@ export class TileViewPage extends React.Component {
+            </div>
+            <Gallery
+              gutter="sm"
+              className="catalog-tile-view-pf catalog-tile-view-pf-no-categories co-catalog-page__grid"
```
Gallery shouldn't have `co-catalog-page__grid` here.

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -782,28 +863,49 @@ export class TileViewPage extends React.Component {
+            ) : (
+              <Gallery
+                gutter="sm"
+                className="catalog-tile-view-pf catalog-tile-view-pf-no-categories co-catalog-page__grid"
```
Move `co-catalog-page__grid` to a parent `div` so that this case and the grouping case use the same parent with this class.

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -641,6 +686,14 @@ export class TileViewPage extends React.Component {
+      this.onFilterChange('groupBy', null, '');
+    } else {
+      this.onFilterChange('groupBy', null, value);
+    }
```
Why is it not simply `this.onFilterChange('groupBy', null, value);`

Although grouping is not a filter, so you shouldn't be using `onFilterChange`.

debsmita1 replied:
Not using `onFilterChange` anymore

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -641,6 +686,14 @@ export class TileViewPage extends React.Component {
     this.setState({ filterGroupsShowAll: updatedShow });
   }
 
+  onDropdownChange(value) {
```
`onDropdownChange` => `onGroupChange`

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -617,8 +660,10 @@ export class TileViewPage extends React.Component {
+    if (filterType === FilterType.keyword) {
+      updateURLParams(FilterType.keyword, `${value}`);
+    } else if (filterType === FilterType.groupBy) {
+      updateURLParams(FilterType.groupBy, `${value}`);
```
Grouping is not a filter and should be handled separately.

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -310,10 +329,13 @@ const determineAvailableFilters = (initialFilters, items, filterGroups) => {
   activeFilters.keyword.active = !!keywordFilter;
 
+  activeFilters.groupBy.value = groupByFilter || '';
+  activeFilters.groupBy.active = !!groupByFilter;
```
I'm confused. Group by does not actually filter anything, it only affects the visual appearance of the items.
This should not be mixed in with the filters on the page.

debsmita1 replied:
@christianvogt Initially the ask was to display only the Installed Operators in groups. Later on when we decided to add the Non operators grouped together I should have updated the variable names which I missed

debsmita1 replied:
moved it out of activeFilters

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -324,9 +346,12 @@ const getActiveFilters = (keywordFilter, groupFilters, activeFilters) => {
     _.set(activeFilters, 'keyword.active', !!value);
+  } else if (filterType === FilterType.groupBy) {
+    _.set(activeFilters, 'groupBy.value', value);
+    _.set(activeFilters, 'groupBy.active', !!value);
```
Grouping is not a filter.

benjaminapetersen commented:
/assign @christianvogt 

I believe.

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -21,9 +21,21 @@ import {
+};
+
+export const groupByTypes = {
+  Operator: 'Operator',
```
I just realized that `tile-view-page` is a generic component. You must parameterize the grouping strategy using props and control it from the parent.

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -227,6 +238,17 @@ const filterByGroup = (items, filters) => {
   );
 };
 
+export const groupItems = (items, groupBy) => {
```
This specific grouping needs to be handled by the parent component `catalog-items.jsx`

christianvogt commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -764,9 +832,39 @@ export class TileViewPage extends React.Component {
+              className="catalog-tile-view-pf catalog-tile-view-pf-no-categories"
+            >
+              {_.map(value, (item) => this.renderGalleryItem(item, renderTile))}
+            </Gallery>
```
This is duplicate code with that of the render function.

divyanshiGupta commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -323,8 +345,16 @@ const getActiveFilters = (keywordFilter, groupFilters, activeFilters) => {
   return activeFilters;
 };
 
+export const updateActiveDisplayType = (activeDisplayType, displayType, value) => {
```
Add types for the parameters

debsmita1 replied:
It's a jsx file

divyanshiGupta replied:
Oh missed that. Ignore this comment then :P

divyanshiGupta commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -399,13 +429,13 @@ const setURLParams = (params) => {
 };
 
-export const updateURLParams = (filterName, value) => {
+export const updateURLParams = (paramName, value) => {
```
Add types

debsmita1 replied:
It's a jsx file

divyanshiGupta replied:
Oh missed that. Ignore this comment then :P

rohitkrai03 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -591,34 +622,40 @@ export class TileViewPage extends React.Component {
     const clearedFilters = clearActiveFilters(activeFilters, filterGroups);
 
-    this.updateMountedState(this.getUpdatedState(categories, selectedCategoryId, clearedFilters));
+    this.updateMountedState(
```
You can remove all the extra logic for grouping from the filter logic.

debsmita1 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -323,8 +329,16 @@ const getActiveFilters = (keywordFilter, groupFilters, activeFilters) => {
 };
 
+export const updateActiveDisplayType = (activeDisplayType, displayType, value) => {
+  if (displayType === DisplayTypes.groupBy) {
```
Removed DisplayTypes

debsmita1 commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -546,12 +571,17 @@ export class TileViewPage extends React.Component {
   }
 
-  getUpdatedState(categories, selectedCategoryId, activeFilters) {
+  getUpdatedState(categories, selectedCategoryId, activeFilters, activeDisplayType) {
```
removed `groupBy` strategy from here

rohitkrai03 commented:
/assign @christianvogt

debsmita1 commented:
/test e2e-gcp-console

debsmita1 commented:
/test analyze

andrewballantyne commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -764,9 +773,38 @@ export class TileViewPage extends React.Component {
+            </Title>
+            {this.renderItems(value, renderTile)}
+            <br />
+          </>
```
I think this can be better represented with css rather than a `br`. I think `br`s are old fashioned and we lose the fidelity to control the gap. 

Also, you lack the `key` prop. Anytime you do `map` you need to use the `key` prop on returned JSX.

debsmita1 replied:
- added key prop 
- removed <br/> and added a css class to handle the space between the groups

andrewballantyne commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -467,10 +468,12 @@ const defaultFilters = {
   },
 };
 
+const none = 'None';
```
Not quite what I meant :) For one, this will fail the linter as it's defined after its use... For two, this won't help using two different variables to match. You either need to use this exclusively... or use the _enum_ we talked about.

debsmita1 replied:
Sorry for that. Now using `groupByTypes.None` everywhere instead of `'None'`

andrewballantyne commented:
/hold

e2e tests are still failing, no need for the bot to hammer after it fails.

andrewballantyne commented:
/kind feature

christianvogt commented:
/approve

spadgett commented:
/hold cancel
/retest
#3847 has fixed the e2e tests
